### PR TITLE
sharing SerializationInfo for corert and coreclr

### PR DESCRIPTION
**Pull request by maryamariyan**
I tested locally on corert and on corefx and ran tests to make sure there is nothing is breaking.

I did multiple commits so that it's easier to review. Overall there was not much implementation difference more so it was code styles or renamings in the diff than anything else.

I had to add CORECLR condition in two places only.

I added `_requireSameTokenInPartialTrust` as it was missing in one of the repos. But when you trace its usages in this class you see it doesn't really do much. So perhaps if agreed I can remove it instead.

There is a noticeable summary comment on top of `UpdateValue` method and seems like the api is only public for corefx access. Let me know if we would like to keep it public. It was public in both repos anyway. I did however end up merging the comment contents for this method.

cc: @jkotas @danmosemsft

### CODE REVIEW

**Comment by danmosemsft** [comment]
> ran tests 

I assume you mean on CoreCLR, not CoreRT (I actually don't know how to run CoreFX tests on CoreRT although certainly they work)

File: src/System.Private.CoreLib/src/System/Runtime/Serialization/SerializationInfo.cs [hunk 1]
```diff
@@ -2,46 +2,28 @@
...
-        private Dictionary<string, int> m_nameToIndex;
-        internal int m_currMember;
-        internal IFormatterConverter m_converter;
-        private String m_fullTypeName;
-        private String m_assemName;
-        private Type objectType;
-        private bool isFullTypeNameSetExplicit;
-        private bool isAssemblyNameSetExplicit;
-        private bool requireSameTokenInPartialTrust;
+        private string[] _names;
```
Comment (danmosemsft) [inline src/System.Private.CoreLib/src/System/Runtime/Serialization/SerializationInfo.cs]: @jkotas do you know why some of these were internal?

**Comment by maryamariyan** [comment]
@danmosemsft @jkotas please also take a look at this corert PR to see how the diff will look like in there https://github.com/dotnet/corert/pull/5847

The two files will end up looking identical. What would help with looking at the diff per commit is to see how I ended up converging them into the same content

**Comment by maryamariyan** [comment]
>> ran tests 

>I assume you mean on CoreCLR, not CoreRT (I actually don't know how to run CoreFX tests on CoreRT although certainly they work)

@danmosemsft when `build.cmd` finishes running there is a log that states some number of tests passed. That's all I did for corert. But for corefx I followed instructions to build corefx using local coreclr package and ran tests to make sure there is no failures.

File: src/System.Private.CoreLib/src/System/Runtime/Serialization/SerializationInfo.cs (hunk 1 above)
Comment (jkotas) [reply src/System.Private.CoreLib/src/System/Runtime/Serialization/SerializationInfo.cs]: No idea. It is not unusual for the internal/private in the old code like this to be non-nonsensical.

File: src/System.Private.CoreLib/shared/System/Runtime/Serialization/SerializationInfo.cs [hunk 2]
```diff
@@ -0,0 +1,514 @@
...
+        private Type[] _types;
+        private int _count;
+        private Dictionary<string, int> _nameToIndex;
+        private IFormatterConverter _converter;
+        private string _rootTypeName;
+        private string _rootTypeAssemblyName;
+        private Type _rootType;
+
+        [CLSCompliant(false)]
+        public SerializationInfo(Type type, IFormatterConverter converter)
```
Comment (jkotas) [inline src/System.Private.CoreLib/shared/System/Runtime/Serialization/SerializationInfo.cs]: Nit: This is not the CoreRT version ... the CoreRT version is:

```
        [CLSCompliant(false)]
        public SerializationInfo(Type type, IFormatterConverter converter, bool requireSameTokenInPartialTrust)
            : this(type, converter)
        {
            // requireSameTokenInPartialTrust is a vacuous parameter in a platform that does not support partial trust.
        }
```

**Comment by jkotas** [comment]
@dotnet-bot test Windows_NT x64 Checked corefx_baseline

File: src/System.Private.CoreLib/shared/System/Runtime/Serialization/SerializationInfo.cs (hunk 2 above)
Comment (maryamariyan) [reply src/System.Private.CoreLib/shared/System/Runtime/Serialization/SerializationInfo.cs]: I just moved this lower down on line 53

**Comment by maryamariyan** [comment]
@dotnet-bot test Windows_NT x64 Checked corefx_baseline

**Comment by maryamariyan** [comment]
@dotnet-bot test Windows_NT x64 Checked corefx_baseline

**Pull request merged**